Storage: Add STP for offline storage migration (#76)
RedHatQE/openshift-virtualization-tests-design-docs: stps/sig-storage/storage_mig_offline.md

# Openshift-virtualization-tests Test plan

## **Offline Storage Migration - Quality Engineering Plan**

### **Metadata & Tracking**

> - **Enhancement(s):** https://github.com/kubevirt/kubevirt-migration-controller/pull/32

Reviewer: is there a vep for this feature? if so, please link to it and not to the code changes

Reviewer 2: No veps for this feature.

- **Feature Tracking:** [CNV-73509](https://redhat.atlassian.net/browse/CNV-73509)
- **Epic Tracking:** [CNV-73500](https://redhat.atlassian.net/browse/CNV-73500)
- **QE Owner(s):** Jose Manuel Castano (joscasta@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage

> 

Reviewer: please add feature maturity (was recently added to the template)

Author: Done in next commit!

**Document Conventions:**
None

### **Feature Overview**

This feature extends the OpenShift Virtualization migration plan to support storage migration for offline (stopped) VMs in addition to existing online (running) VM support. It enables customers to migrate storage for VMs regardless of their running state, allowing mixed migration plans containing both offline and running VMs.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**

>   - *List the key D/S requirements reviewed:* Reviewed the user cases for offline VM storage migration from CNV-82430 and CNV-73500

Reviewer: please list the requirements.

Author: Addressed in next commit!

- [x] **Understand Value and Customer Use Cases**
  - *Describe the feature's value to customers:* Customers need to perform storage migration for offline VMs without requiring them to be running, providing flexibility in storage management operations.
  - *List the customer use cases identified:* Storage migration for mixed offline VMs and running VMs in one migration plan, allowing batch migration operations regardless of VM state.

- [x] **Testability**
  - *Note any requirements that are unclear or untestable:* Requirements are testable. Downstream build with the feature code is available for testing.

- [x] **Acceptance Criteria**
  - *List the acceptance criteria:*
    - Storage migration completes successfully for offline VMs between ODF and HPP storage classes

>     - Storage migration completes successfully for mixed offline VMs and running VMs

Reviewer: what does `completes successfully` mean? acceptance criteria needs to include measurables

Author: I modify it in next commit

    - Storage migration completes successfully for offline VMs with hotplug disk
    - Source volume could be retained/cleaned up for an offline VM migration completed with retentionPolicy defined
    - Offline VM points to the origin volume when migration failed
    - Migration succeeded when starting a stopped VM during migration

  - *Note any gaps or missing criteria:* N/A

- [x] **Non-Functional Requirements (NFRs)**
  - *List applicable NFRs and their targets:* Documentation updates to reflect offline VM storage migration support and UI support for offline VM migrations.

>   - *Note any NFRs not covered and why:* Performance, Monitoring, Observability, Security and Scalability testing are not included in this test plan

Reviewer: please indicate, for each nfr, **why** it is not needed / covered

Author: Added the **why** in next commit.

  - *Testing impact:* No API test updates required; functional tests will verify new behavior

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*

- [x] **Topology Considerations**
  - *Describe topology requirements:* Standard 3-master/3-worker cluster sufficient
  - *Impact on test design:* No special topology requirements

### **II. Software Test Plan (STP)**

This STP serves as the **overall roadmap for testing**, detailing the scope, approach, resources, and schedule.

#### **1. Scope of Testing**

**Testing Goals**

- **[P0]** Verify offline VM storage migration completes between ODF and HPP storage classes, and the VM boots successfully after migration
- **[P0]** Verify storage migration completes for a migration plan containing both offline and running VMs
- **[P0]** Verify source volumes are retained or deleted according to retentionPolicy configuration when offline VM storage migration completes
- **[P1]** Verify offline VM storage migration completes when the VM has hotplug disks attached
- **[P2]** Verify offline VM continues pointing to the original volume when storage migration fails

> - **[P2]** Verify storage migration completes when a stopped VM is started during the migration process

Reviewer 3: I'm not sure how we can reliably automate this test, storage migration may happen really fast, and we may not be able to start the VM in time

Author: The only thing that comes to my mind would be use large DVs, this may make the migration takes longer time to start/stop the VM, if this is possible.

Reviewer 2: There is no way to limit the migration speed using CDI, so the only thing would be to use larger images that take a while to complete. Like a windows image or something like that.

**Storage Class Coverage**

The following storage class migration combinations will be tested:

> - **ODF** (ocs-storagecluster-ceph-rbd-virtualization) ↔ **HPP** (hostpath-provisioner)

Reviewer 3: ```suggestion
- **ODF** (ocs-storagecluster-ceph-rbd-virtualization) ↔ **HPP** (hostpath-csi-pvc-block)
```

Author: Changed in next commit

- **ODF ↔ ODF** — Same storage class migration
- **HPP ↔ HPP** — Same storage class migration

Storage classes **not covered** in this test plan:

> - Cloud provider-specific storage classes (AWS EBS, Azure Disk, GCP PD) — Out of scope for initial release

Reviewer: should go under `out of scope` and have PM sign off

Author: Moved into `out of scope` section. PM sign off I will added once reviewed by PM

- [ ] **Upgrade Testing** — Validates upgrade paths from previous versions, data migration, and configuration preservation
  - *Details:* Not applicable

- [ ] **Dependencies** — Blocked by deliverables from other components/products. Identify what we need from other teams before we can test.
  - *Details:* No blocking dependencies

- [x] **Cross Integrations** — Does the feature affect other features or require testing by other teams? Identify the impact we cause.
  - *Details:* UI team needs to update their migration UI to support offline VM selection

**Infrastructure**

- [ ] **Cloud Testing** — Does the feature require multi-cloud platform testing? Consider cloud-specific features.
  - *Details:* Not applicable

#### **3. Test Environment**

- **Cluster Topology:** 3-master/3-worker bare-metal

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization 4.22

- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM

- **Special Hardware:** N/A

> - **Storage:** ocs-storagecluster-ceph-rbd-virtualization, hostpath-provisioner

Reviewer 3: ```suggestion
- **Storage:** ocs-storagecluster-ceph-rbd-virtualization, hostpath-csi-pvc-block
```

Author: Changed in next commit

- **Network:** OVN-Kubernetes, IPv4

- **Required Operators:** N/A

- **Platform:** PSI

- **Special Configurations:** N/A

#### **3.1. Testing Tools & Frameworks**

- **Test Framework:** Standard

- **CI/CD:** N/A

- **Other Tools:** N/A

#### **4. Entry Criteria**

The following conditions must be met before testing can begin:

- [x] Requirements and design documents are **approved and merged**
- [x] Test environment can be **set up and configured** (see Section II.3 - Test Environment)

#### **5. Exit Criteria**

- [ ] All high-priority defects are resolved and verified
- [ ] Test coverage goals achieved
- [ ] Test automation merged (required for GA sign-off)
- [ ] All planned test cycles completed
- [ ] Test summary report approved

> - [ ] Acceptance criteria met

Reviewer: please remove; was removed from the template a while back

Author: Removed in next commit

> #### **6. Risks**

Reviewer: please see https://github.com/RedHatQE/openshift-virtualization-tests-design-docs/pull/65 how to fill this section;

Author: Addressed in next commit, following template.

# Openshift-virtualization-tests Test plan

## **Offline Storage Migration - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** [CNV-77501](https://redhat.atlassian.net/browse/CNV-77501) - no VEP exists for this feature
- **Feature Tracking:** [CNV-73509](https://redhat.atlassian.net/browse/CNV-73509)
- **Epic Tracking:** [CNV-73500](https://redhat.atlassian.net/browse/CNV-73500)
- **QE Owner(s):** Jose Manuel Castano (joscasta@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage
- **Feature Maturity:**
  - TP: v4.22

>   - GA: v5.0

Author: Got it, done in next commit!

- **[P0]** Verify source volumes are retained or deleted according to retentionPolicy configuration when offline VM storage migration completes
- **[P1]** Verify offline VM storage migration completes when the VM has hotplug disks attached
- **[P2]** Verify offline VM continues pointing to the original volume when storage migration fails
- **[P2]** Verify storage migration completes when a stopped VM is started during the migration process

**Storage Class Coverage**

The following storage class migration combinations will be tested:
- **ODF** (ocs-storagecluster-ceph-rbd-virtualization) ↔ **HPP** (hostpath-csi-pvc-block)
- **ODF ↔ ODF** — Same storage class migration
- **HPP ↔ HPP** — Same storage class migration

**Out of Scope (Testing Scope Exclusions)**

The following items are explicitly Out of Scope for this test cycle and represent intentional exclusions.
No verification activities will be performed for these items, and any related issues found will not be classified as defects for this release.

- **Storage Classes:** Cloud provider-specific storage classes (AWS EBS, Azure Disk, GCP PD) are out of scope for initial release

> > **PM Sign-off:** [Name], [Date]

Author: I will add it once I get approval from PM

  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g., status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g., status conditions, events, or notifications indicating success or failure)
  - *Details:* UI testing will be covered in https://redhat.atlassian.net/browse/CNV-77503

- [ ] **Monitoring** — Does the feature require metrics and/or alerts?
  - *Details:* Not applicable

**Integration & Compatibility**

- [x] **Compatibility Testing** — Ensures feature works across supported platforms, versions, and configurations
  - Does the feature maintain backward compatibility with previous API versions and configurations?
  - *Details:* Feature maintains backward compatibility with existing migration API. Existing online VM migrations continue to work unchanged.

- [ ] **Upgrade Testing** — Validates upgrade paths from previous versions, data migration, and configuration preservation
  - *Details:* Not applicable

> 

Author: Changed in next commit

- [ ] **Dependencies** — Blocked by deliverables from other components/products. Identify what we need from other teams before we can test.
  - *Details:* No blocking dependencies

- [x] **Cross Integrations** — Does the feature affect other features or require testing by other teams? Identify the impact we cause.
  - *Details:* UI team needs to update their migration UI to support offline VM selection

**Infrastructure**

- [ ] **Cloud Testing** — Does the feature require multi-cloud platform testing? Consider cloud-specific features.
  - *Details:* Not applicable

#### **3. Test Environment**

> - **Cluster Topology:** 3-master/3-worker bare-metal

Author: Yes you're right, both should coincide. Updated in next commit

- **Feature Tracking:** [CNV-73509](https://redhat.atlassian.net/browse/CNV-73509)
- **Epic Tracking:** [CNV-73500](https://redhat.atlassian.net/browse/CNV-73500)
- **QE Owner(s):** Jose Manuel Castano (joscasta@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage
- **Feature Maturity:**
  - DP: N/A
  - TP: v4.22
  - GA: v5.0

**Document Conventions:**
None

### **Feature Overview**

This feature extends the OpenShift Virtualization migration plan to support storage migration for offline (stopped) VMs in addition to existing online (running) VM support. It enables customers to migrate storage for VMs regardless of their running state, allowing mixed migration plans containing both offline and running VMs.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**
  - *List the key D/S requirements reviewed:*

>     - Support storage migration for offline (stopped) VMs between different storage classes

Reviewer 4: will you cover between File and Block, (both directions)? If so, please call it out.

Reviewer 2: We should support all the combinations since we are just calling a CDI copy clone to do the actual transfer.

Author: You are right, should cover all volume mode combinations. Let me add it in next commit.

    - Support mixed migration plans containing both offline and running VMs
    - Support offline VM storage migration with hotplug disks attached
    - Support retentionPolicy configuration for source volume cleanup after offline VM migration
    - Ensure offline VMs remain pointing to original volumes when migration fails

>     - Support VM start operations during ongoing storage migration without conflicts

Reviewer 2: The main point here being that once you have started a migration, we changed the volumes to the new volume (which will still be in pending until the transfer completes) and thus if you start the VM during migration, the VM will not start until the migration completes.

Author: Then I understand that if the VM restarts for some reason, will happen the same and the VM will not start, right?
Anyway I see this as a good test case, but needs more explanation to clarify that the VM only will start when migration completes.

Reviewer 2: So when the migration starts when the VM is offline, we first change the volume the VM is using to the new volume, and then start the copy. As a result now if someone starts the VM, the VM will remain pending because the DataVolume is not in succeeded phase.

Author: Modified to reflect test case scenario where the VM starts during migration and it should complete once migration completes.

- **[P0]** Verify storage migration completes for a migration plan containing both offline and running VMs
- **[P0]** Verify source volumes are retained or deleted according to retentionPolicy configuration when offline VM storage migration completes
- **[P1]** Verify offline VM storage migration completes when the VM has hotplug disks attached
- **[P2]** Verify offline VM continues pointing to the original volume when storage migration fails
- **[P2]** Verify storage migration completes when a stopped VM is started during the migration process

**Storage Class Coverage**

The following storage class migration combinations will be tested:
- **ODF** (ocs-storagecluster-ceph-rbd-virtualization) ↔ **HPP** (hostpath-csi-pvc-block)
- **ODF ↔ ODF** — Same storage class migration
- **HPP ↔ HPP** — Same storage class migration

**Out of Scope (Testing Scope Exclusions)**

The following items are explicitly Out of Scope for this test cycle and represent intentional exclusions.
No verification activities will be performed for these items, and any related issues found will not be classified as defects for this release.

> - **Storage Classes:** Cloud provider-specific storage classes (AWS EBS, Azure Disk, GCP PD) are out of scope for initial release

Reviewer 4: Not sure why we are excluding cloud storage classes, there should be no difference for something run in the cloud vs/ on-prem. If this is a resource constraint, that's a different conversation.

Reviewer 2: I don't know either, as we are just doing a CDI copy clone it will work with cloud providers as well. But Jose would know why he excluded this in particular

Author: Miss understood this, thanks for the clarification. I update the storage classes to include cloud ones. Please let me know if this approach looks correct for you or if you consider a different way to structure the cloud storage class coverage.

# Openshift-virtualization-tests Test plan

## **Offline Storage Migration - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** [CNV-77501](https://redhat.atlassian.net/browse/CNV-77501) - no VEP exists for this feature
- **Feature Tracking:** [CNV-73509](https://redhat.atlassian.net/browse/CNV-73509)
- **Epic Tracking:** [CNV-73500](https://redhat.atlassian.net/browse/CNV-73500)
- **QE Owner(s):** Jose Manuel Castano (joscasta@redhat.com)
- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage
- **Feature Maturity:**
  - DP: N/A
  - TP: v4.22
  - GA: v5.0

**Document Conventions:**
None

> 

Reviewer: **[HIGH] Feature Overview describes internal extension instead of user capability**

"extends the OpenShift Virtualization migration plan" describes an internal system change. The Feature Overview should describe what the feature does from the user's perspective.

Proposed rewording:
> This feature enables customers to migrate storage for VMs regardless of their running state, including stopped VMs. Customers can perform batch storage migrations that include both offline and running VMs in a single operation.

Author: Changed to match more user capability description in next commit

### **Feature Overview**

This feature extends the OpenShift Virtualization migration plan to support storage migration for offline (stopped) VMs in addition to existing online (running) VM support. It enables customers to migrate storage for VMs regardless of their running state, allowing mixed migration plans containing both offline and running VMs.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**
  - *List the key D/S requirements reviewed:*

>     - Support storage migration for offline (stopped) VMs between different storage classes

Reviewer: **[HIGH] Requirements include implementation terms — rewrite from user perspective**

Several requirements reference API field names and internal mechanisms:
- `retentionPolicy configuration` → use "source volume cleanup policy" (user-facing term)
- `CDI copy clone performs the underlying storage transfer` → describe in user terms (e.g., "the feature supports all volume mode combinations for cross-storage migration")
- `volumes are switched to new target volumes in pending state once migration begins` → describe the user-observable behavior instead, e.g., "When a user starts a stopped VM during migration, the VM waits for migration to complete before becoming ready — the user sees the VM transition to Running only after storage migration finishes"

STPs should describe what users experience, not how the system works internally.

Author: Rewritten to follow more user perspective in next commit

    - Support all volume mode combinations (File-to-Block, Block-to-File, File-to-File, Block-to-Block) given that CDI copy clone performs the underlying storage transfer
    - Reviewed user cases for offline VM storage migration from CNV-82430 and CNV-73500

- [x] **Understand Value and Customer Use Cases**
  - *Describe the feature's value to customers:* Customers need to perform storage migration for offline VMs without requiring them to be running, providing flexibility in storage management operations.
  - *List the customer use cases identified:*
    - As a VM owner, I want to migrate storage for offline VMs, so that I can perform storage management operations without having to start the VMs
    - As a VM owner, I want to migrate storage with mixed VM states (online and offline) in one migration plan, so that I can perform batch migration operations regardless of VM state

- [x] **Testability**
  - *Note any requirements that are unclear or untestable:* Requirements are testable. Downstream build with the feature code is available for testing.

- [x] **Acceptance Criteria**
  - *List the acceptance criteria:*

>     - As a VM owner, I want to migrate storage for offline VMs between ODF and HPP storage classes, so that:

Reviewer: **[CRITICAL] Acceptance criteria don't prove non-disruptive behavior**

The acceptance criteria for "start VM during migration" verify end-state only (VM started, points to target storage). A disrupt-then-restart sequence would produce the same passing result.

Per STP guidelines: acceptance criteria must include at least one condition that can *only* pass if disruption never occurred. For example: "VM does not transition to Running phase during migration" — this would fail if a disrupt-then-restart occurred, proving the pending behavior is real.

Author: Added this other acceptance criteria in next commit.

      - Migration plan status reports "Succeeded"
      - Offline VM disk references point to the new target storage class
      - VM boots successfully after migration using the migrated storage
    - As a VM owner, I want to migrate storage with mixed offline VMs and running VMs in one migration plan, so that:
      - Migration plan status reports "Succeeded"
      - All offline VMs point to target storage class
      - All running VMs point to target storage class and remain running
    - As a VM owner, I want to migrate storage for offline VMs with hotplug disk, so that:
      - Migration plan status reports "Succeeded"

>       - All disks including hotplug disks are migrated to target storage class

Reviewer: **[CRITICAL] Scalability NFR dismissed but feature relies on infrastructure with existing parallelism limits**

The feature relies on CDI copy clone and existing migration infrastructure, both of which have cluster-level parallelism constraints. Saying "Not applicable" is insufficient — existing scale limits must be acknowledged even if the feature introduces no new requirements.

Suggestion: Change to something like: "Scalability characteristics inherit from existing migration infrastructure. CDI clone parallelism and concurrent migration limits apply — no new scalability requirements introduced by offline VM support."

Author: Changed in next commit

- [x] **Non-Functional Requirements (NFRs)**
  - *List applicable NFRs and their targets:* Documentation updates to reflect offline VM storage migration support and UI support for offline VM migrations.
  - *Note any NFRs not covered and why:*
    - **Performance:** Not covered - Performance testing for bulk offline migrations is tracked separately and will be addressed by a dedicated test plan
    - **Monitoring:** Not applicable - Feature does not introduce new metrics or alerts; existing migration monitoring applies
    - **Observability:** Not applicable - Feature reuses existing migration observability patterns without new requirements
    - **Security:** Not applicable - Feature does not introduce new security boundaries or authentication/authorization requirements; leverages existing migration RBAC
    - **Scalability:** Not applicable - Scalability characteristics inherit from existing migration infrastructure; no new scalability requirements introduced

#### **2. Known Limitations**

The limitations are documented to ensure alignment between development, QA, and product teams.
The following topics will not be tested or supported.

> None - reviewed and confirmed with Jose Manuel Castano on Apr 28,2026.

Reviewer: **[HIGH] Technology Challenges section is generic — no actual testing challenges described**

"Offline migration uses a different mechanism than online migration" is vague, and the impact statement just restates the testing goals. This section should list specific challenges and their impact on testing — e.g., how to simulate migration failures for offline VMs, timing issues with VM start during migration, volume mode conversion behavior differences.

Author: Done in next commit

#### **3. Technology and Design Review**

- [x] **Developer Handoff/QE Kickoff**
  - *Key takeaways and concerns:* Extend the offline VMs storage migration support

- [x] **Technology Challenges**
  - *List identified challenges:* Offline migration uses a different mechanism than online migration, requiring separate test coverage
  - *Impact on testing approach:* Test cases must verify both offline VM migration mixed scenarios with both offline and running VMs in the same migration plan.

- [x] **API Extensions**
  - *List new or modified APIs:* No new APIs - extends existing migration plan API to handle offline VMs
  - *Testing impact:* No API test updates required; functional tests will verify new behavior

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*

- [x] **Topology Considerations**
  - *Describe topology requirements:* Standard 3-master/3-worker cluster sufficient

>   - *Impact on test design:* No special topology requirements

Reviewer: **[WARNING] Same-storage-class migration claimed in coverage but no test scenario exists**

The "Storage Class Coverage" section promises same-storage-class migrations (ODF↔ODF, HPP↔HPP) will be tested, but all six test scenarios in Section III reference only cross-storage-class migration ("between ODF and HPP"). Either add a test scenario for same-storage-class migration or remove the claim from the coverage section.

Author: Would make more sense only reflect cross storage class migrations as same storage class migrations wouldn't make much sense. Removed in next commit, thanks

Reviewer 2: There is a use case for same storage class migrations, in particular hpp would make a good candidate. The scenario is you have a number of nodes all of which only have local storage (hpp). You want to live migration to another node. You cannot do this with a compute live migration because you don't have RWX storage. But you can do it with a storage live migration. So hpp -> hpp makes sense for that particular scenario.

Author: Ok, got it, then would be good to have a specif test case for that scenario. I will add it in next commit.

- **[P2]** Verify storage migration completes when a stopped VM is started during the migration process and the VM waits for migration completion before becoming ready

**Storage Class Coverage**

The following storage classes are supported for migration testing:
- **ODF** (ocs-storagecluster-ceph-rbd-virtualization)
- **HPP** (hostpath-csi-pvc-block)

Migration combinations will be tested including:
- Cross-storage class migrations (e.g., ODF ↔ HPP)
- Same-storage class migrations (e.g., ODF ↔ ODF, HPP ↔ HPP)

**Volume Mode Coverage**

All volume mode combinations are supported and will be tested:
- **Block ↔ Block** — Block to Block migration
- **File ↔ File** — Filesystem to Filesystem migration
- **Block ↔ File** — Block to Filesystem migration
- **File ↔ Block** — Filesystem to Block migration

**Out of Scope (Testing Scope Exclusions)**

The following items are explicitly Out of Scope for this test cycle and represent intentional exclusions.
No verification activities will be performed for these items, and any related issues found will not be classified as defects for this release.

None

> 

Reviewer: **[HIGH] Test Limitations subsection is missing**

The Scope of Testing section has "Out of Scope" but no "Test Limitations" subsection. These are distinct required concepts: Test Limitations are constraints imposed on QE, while Out of Scope are decisions QE made.

If no test limitations apply, add:

**Test Limitations**

None — reviewed and confirmed with [Name], [Date].

Author: Added in next commit

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**
  - *List the key D/S requirements reviewed:*
    - Support storage migration for offline (stopped) VMs between different storage classes
    - Support mixed migration plans containing both offline and running VMs
    - Support offline VM storage migration with hotplug disks attached
    - Support source volume cleanup policy configuration for offline VM migration (retain or delete source volumes after migration completes)
    - Ensure offline VMs remain pointing to original volumes when migration fails
    - Support VM start operations during ongoing storage migration — when a user starts a stopped VM during migration, the VM waits for migration to complete before becoming ready (the VM transitions to Running only after storage migration finishes)
    - Support all volume mode combinations for cross-storage migration (File-to-Block, Block-to-File, File-to-File, Block-to-Block)

>     - Reviewed user cases for offline VM storage migration from CNV-82430 and CNV-73500

Reviewer: **[WARNING] Requirements list includes a review note, not a requirement**

Line 43 ("Reviewed user cases for offline VM storage migration from CNV-82430 and CNV-73500") is a QE review note, not a testable D/S requirement. Remove it from the requirements list — the Jira references belong in metadata or review checklist context, not as a requirement item.

Author: Removed in next commit

      - All running VMs point to target storage class and remain running
    - As a VM owner, I want to migrate storage for offline VMs with hotplug disk, so that:
      - Migration plan status reports "Succeeded"
      - All disks including hotplug disks are migrated to target storage class
      - VM boots successfully with all disks accessible
    - As a VM owner, I want to retain or delete the source volume for an offline VM, so that:
      - When the cleanup policy is set to retain: source volumes exist after migration completes
      - When the cleanup policy is set to delete: source volumes are deleted after migration completes
    - As a VM owner, I want an offline VM to point to the original volume when migration fails, so that:
      - Migration plan status reports "Failed"
      - VM disk references remain unchanged and point to original storage
    - As a VM owner, I want migration to succeed when starting a stopped VM during migration, so that:
      - Migration plan status reports "Succeeded"
      - VM remains in a pending state throughout the migration and only becomes Ready after migration completes (proves no disrupt-then-restart occurred)

>       - VM start operation waits for migration to complete before the VM becomes ready (volumes are switched to target storage in pending state)

Reviewer: **[WARNING] Acceptance criterion uses internal implementation language**

Line 77 ("volumes are switched to target storage in pending state") describes internal system behavior. The user-observable criterion on line 76 already proves non-disruptive behavior — line 77 is redundant and violates the user-perspective rule.

```suggestion
      - VM start operation waits for migration to complete before the VM becomes ready
```

Author: Done in next commit

      - VM points to target storage and starts successfully after migration completes

  - *Note any gaps or missing criteria:* N/A

- [x] **Non-Functional Requirements (NFRs)**

>   - *List applicable NFRs and their targets:* Documentation updates to reflect offline VM storage migration support and UI support for offline VM migrations.

Reviewer: **[CRITICAL] UI and Documentation NFRs need explicit per-category entries**

The NFR section must explicitly address all seven categories (Monitoring, Observability, UI, Documentation, Performance, Security, Scalability). UI and Documentation are mentioned only in the summary line (83) without per-category targets or ownership.

Suggestion — split into applicable vs not-covered lists:

```suggestion
  - *List applicable NFRs and their targets:*
    - **Documentation:** User guide updates for offline VM storage migration support
    - **UI:** UI updates for offline VM migration selection (covered by UI team in CNV-77503)
  - *Note any NFRs not covered and why:*
    - **Performance:** Not covered - Performance testing for bulk offline migrations is tracked separately and will be addressed by a dedicated test plan
    - **Monitoring:** Not applicable - Feature does not introduce new metrics or alerts; existing migration monitoring applies
    - **Observability:** Not applicable - Feature reuses existing migration observability patterns without new requirements
    - **Security:** Not applicable - Feature does not introduce new security boundaries or authentication/authorization requirements; leverages existing migration RBAC
    - **Scalability:** Inherits existing limits - Feature reuses existing migration infrastructure with its cluster-level parallelism constraints; offline VM support introduces no additional scalability requirements
```

Author: Fixed in next commit

  - *Note any NFRs not covered and why:*
    - **Performance:** Not covered - Performance testing for bulk offline migrations is tracked separately and will be addressed by a dedicated test plan
    - **Monitoring:** Not applicable - Feature does not introduce new metrics or alerts; existing migration monitoring applies
    - **Observability:** Not applicable - Feature reuses existing migration observability patterns without new requirements
    - **Security:** Not applicable - Feature does not introduce new security boundaries or authentication/authorization requirements; leverages existing migration RBAC

>     - **Scalability:** Inherits existing limits - Feature reuses existing migration infrastructure with its cluster-level parallelism constraints (CDI clone concurrency and migration plan limits); offline VM support introduces no additional scalability requirements

Reviewer: **[WARNING] Internal component names in NFR and Cloud Testing sections**

Lines 89 and 228 reference "CDI clone concurrency" and "CDI copy clone" — internal implementation terms. Rewrite in user-facing language:

Line 89 suggestion:
```suggestion
    - **Scalability:** Inherits existing limits - Feature reuses existing migration infrastructure with its cluster-level parallelism constraints; offline VM support introduces no additional scalability requirements
```

Line 228 suggestion:
```suggestion
  - *Details:* Cloud storage migration uses platform default storage classes and does not require dedicated cloud-specific test scenarios. The underlying storage transfer mechanism is platform-agnostic. Testing with bare-metal storage classes (ODF and HPP) validates the migration mechanism; cloud platforms (AWS EBS, Azure Disk, GCP PD) inherit this validated behavior through the same infrastructure.
```

Author: Done in next commit

- [x] Test environment can be **set up and configured** (see Section II.3 - Test Environment)

#### **5. Risks**

**Timeline/Schedule**

- **Risk:** No scheduling or deadline risks identified
  - **Mitigation:** Standard test timeline is sufficient for planned test scenarios

**Test Coverage**

- **Risk:** No gaps in test coverage identified
  - **Mitigation:** All acceptance criteria are covered by planned test scenarios

**Test Environment**

- **Risk:** No hardware, software, or infrastructure constraints identified
  - **Mitigation:** Standard test environment is sufficient for testing this feature

**Untestable Aspects**

- **Risk:** No untestable scenarios identified
  - **Mitigation:** All scenarios can be reproduced in test environment

**Resource Constraints**

- **Risk:** No staffing, skill, or capacity limitations identified
  - **Mitigation:** Current QE team capacity is sufficient for planned test execution

**Dependencies**

- **Risk:** No blocking external dependencies identified
  - **Mitigation:** UI team updates for offline VM selection are non-blocking; API testing can proceed independently

---

### **III. Test Scenarios & Traceability**

> - **[CNV-73500]** — As a VM owner, I want to migrate storage for offline VMs between ODF and HPP

Reviewer: **[WARNING] Test scenarios use epic ID instead of requirement ID**

All seven scenarios in Section III map to `[CNV-73500]` (Epic Tracking). Per STP guidelines, scenarios must trace to requirement IDs, not epics. Use `[CNV-77501]` (the enhancement/QE ticket from metadata) or specific D/S requirement tickets where available.

Example fix for Scenario 1:
```suggestion
- **[CNV-77501]** — As a VM owner, I want to migrate storage for offline VMs between ODF and HPP
```

Author: Changed for all test scenarios.

    - As a VM owner, I want to migrate storage for offline VMs between ODF and HPP storage classes, so that:
      - Migration plan status reports "Succeeded"
      - Offline VM disk references point to the new target storage class
      - VM boots successfully after migration using the migrated storage
    - As a VM owner, I want to migrate storage with mixed offline VMs and running VMs in one migration plan, so that:
      - Migration plan status reports "Succeeded"
      - All offline VMs point to target storage class
      - All running VMs point to target storage class and remain running
    - As a VM owner, I want to migrate storage for offline VMs with hotplug disk, so that:
      - Migration plan status reports "Succeeded"
      - All disks including hotplug disks are migrated to target storage class
      - VM boots successfully with all disks accessible
    - As a VM owner, I want to retain or delete the source volume for an offline VM, so that:
      - When the cleanup policy is set to retain: source volumes exist after migration completes
      - When the cleanup policy is set to delete: source volumes are deleted after migration completes

>     - As a VM owner, I want an offline VM to point to the original volume when migration fails, so that:

Reviewer: **[WARNING] Cite existing failure+delete coverage from cleanup STP**

This STP covers cleanup policy retain/delete on success, and failure AC checks that the VM still points at original storage. The stronger invariant — source volumes must still exist when cleanup policy is **delete** and migration fails — is already owned by `stps/sig-storage/storage_mig_cleanup.md`.

Please add one line (e.g. under the failure Testing Goal, Acceptance Criteria, or Out of Scope / existing coverage) so that coverage is explicit:

```suggestion
    - As a VM owner, I want an offline VM to point to the original volume when migration fails, so that:
      - Migration plan status reports "Failed"
      - VM disk references remain unchanged and point to original storage
      - Source-volume survival when cleanup policy is delete and migration fails is covered by stps/sig-storage/storage_mig_cleanup.md
```

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high-fast)*

Author: Addressed into next commit

    - As a VM owner, I want an offline VM to point to the original volume when migration fails, so that:
      - Migration plan status reports "Failed"
      - VM disk references remain unchanged and point to original storage
    - As a VM owner, I want migration to succeed when starting a stopped VM during migration, so that:
      - Migration plan status reports "Succeeded"
      - VM start operation waits for migration to complete before the VM becomes ready (volumes are switched to target storage in pending state)
      - VM points to target storage and starts successfully after migration completes

  - *Note any gaps or missing criteria:* N/A

- [x] **Non-Functional Requirements (NFRs)**
  - *List applicable NFRs and their targets:* Documentation updates to reflect offline VM storage migration support and UI support for offline VM migrations.
  - *Note any NFRs not covered and why:*
    - **Performance:** Not covered - Performance testing for bulk offline migrations is tracked separately and will be addressed by a dedicated test plan
    - **Monitoring:** Not applicable - Feature does not introduce new metrics or alerts; existing migration monitoring applies

>     - **Observability:** Not applicable - Feature reuses existing migration observability patterns without new requirements

Reviewer 2: This is a good point, offline migrations use CDI to do the migration, CDI generates progress information that should be reflected in the migration the same way that live migration should generate progress information.

Author: Modified in next commit, thanks

  - *Impact on testing approach:*
    - Test framework must support failure injection mechanisms for storage operations
    - Test cases need timing coordination to trigger VM start at specific migration stages
    - Verification logic must include data integrity checks across volume mode conversions
    - Mixed migration scenarios require monitoring both offline and online VMs throughout the migration lifecycle

- [x] **API Extensions**
  - *List new or modified APIs:* No new APIs - extends existing migration plan API to handle offline VMs
  - *Testing impact:* No API test updates required; functional tests will verify new behavior

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*

- [x] **Topology Considerations**
  - *Describe topology requirements:* Standard 3-master/3-worker cluster sufficient
  - *Impact on test design:* No special topology requirements

### **II. Software Test Plan (STP)**

This STP serves as the **overall roadmap for testing**, detailing the scope, approach, resources, and schedule.

#### **1. Scope of Testing**

**Testing Goals**

> - **[P0]** Verify offline VM storage migration completes between ODF and HPP storage classes, and the VM boots successfully after migration

Reviewer: **[WARNING] Negative/failure-path goals are P2 — please explain why**

The three P0 goals are happy-path only; the migration-failure goal/scenario is P2. AGENTS.md normally expects a negative path per P0 (or Out of Scope with PM/Lead agreement).

If P2 is intentional, add a short rationale (e.g. under Testing Goals or Out of Scope) explaining why failure-path coverage does not block GA for this feature.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high-fast)*

Author: I'll add a failure-path note to clarify this. Addressed in next commit

Author: Added also a PM agreement line to confirm that Peter agree with this. 
cc @peterclauterbach @aglitke

- [ ] **Security Testing** — Verifies security requirements, RBAC, authentication, authorization, and vulnerability scanning
  - *Details:* N/A. Feature does not introduce new security boundaries or authentication/authorization requirements; leverages existing migration RBAC model.

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g., status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g., status conditions, events, or notifications indicating success or failure)
  - *Details:* UI testing will be covered in https://redhat.atlassian.net/browse/CNV-77503

- [ ] **Monitoring** — Does the feature require metrics and/or alerts?

>   - *Details:* N/A. No new metrics or alerts are required. Feature does not introduce new metrics or alerts; existing migration monitoring applies to both online and offline VM migrations.

Reviewer 2: See my other note, no 'new' metrics are required as the progress field already exists, but we need to verify that offline migrations are hooked up.

Author: I have modified this to include progress reports verification, let me know if you see something wrong

- [ ] **Security Testing** — Verifies security requirements, RBAC, authentication, authorization, and vulnerability scanning
  - *Details:* N/A. Feature does not introduce new security boundaries or authentication/authorization requirements; leverages existing migration RBAC model.

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g., status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g., status conditions, events, or notifications indicating success or failure)
  - *Details:* UI testing will be covered in https://redhat.atlassian.net/browse/CNV-77503

- [x] **Monitoring** — Does the feature require metrics and/or alerts?

>   - *Details:* No new metrics or alerts are required. Existing migration monitoring (including progress field) applies to both online and offline VM migrations. Progress reporting for offline migrations must be verified to ensure it functions correctly.

Reviewer: **[WARNING] Monitoring claims progress verification "must be verified" but no Testing Goal or scenario covers it**

This section states: "Progress reporting for offline migrations must be verified to ensure it functions correctly." However, there is no corresponding Testing Goal in Section II.1 and no test scenario in Section III for progress reporting.

If progress reporting must be verified, please add a Testing Goal and matching scenario. If it's covered by existing regression suites, note that here instead of saying it "must be verified."

Also, "progress field" is an internal API field name — consider rephrasing as "migration progress reporting" for user-facing language.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

Author: Removed the "must be verified" as this is not required in this features. This will be addressed by existing regression suites when time comes as, particularly the migration progress report is still on development.

#### **5. Risks**

**Timeline/Schedule**

- **Risk:** No scheduling or deadline risks identified
  - **Mitigation:** Standard test timeline is sufficient for planned test scenarios

**Test Coverage**

- **Risk:** No gaps in test coverage identified
  - **Mitigation:** All acceptance criteria are covered by planned test scenarios

**Test Environment**

- **Risk:** No hardware, software, or infrastructure constraints identified
  - **Mitigation:** Standard test environment is sufficient for testing this feature

**Untestable Aspects**

- **Risk:** No untestable scenarios identified
  - **Mitigation:** All scenarios can be reproduced in test environment

**Resource Constraints**

- **Risk:** No staffing, skill, or capacity limitations identified
  - **Mitigation:** Current QE team capacity is sufficient for planned test execution

**Dependencies**

- **Risk:** No blocking external dependencies identified
  - **Mitigation:** UI team updates for offline VM selection are non-blocking; API testing can proceed independently

---

### **III. Test Scenarios & Traceability**

- **[CNV-77501]** — As a VM owner, I want to migrate storage for offline VMs between ODF and HPP

>   - *Test Scenario:* [Tier 2] Verify storage migration completes for offline VMs between ODF and HPP across all volume mode combinations (Block-to-Block, File-to-File, Block-to-File, File-to-Block), and the VM boots successfully after migration

Reviewer: **[WARNING] Volume-mode scenario only requires boot, not data integrity**

Technology Challenges correctly call out data-integrity verification across Block↔File conversion, but the Section III scenario only requires migration success and VM boot. Boot-only can pass with inaccessible/corrupted guest data after mode conversion.

**Fix:** Require pre-written guest data remains readable/unchanged after each volume-mode combination.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high-fast)*

Author: Addressed into next commit

  - *Priority:* P0

- **[CNV-77501]** — As a VM owner, I want to migrate storage with mixed VM states (online and offline)

>   - *Test Scenario:* [Tier 2] Verify storage migration completes for a migration plan containing both offline and running VMs

Reviewer: **[WARNING] Several Section III scenarios lack observable pass criteria**

Mixed migration / hotplug / cleanup / failure scenarios say "completes" or "completes successfully" without defining observables. Acceptance Criteria already define them — scenarios should mirror plan status, disk references, running-VM continuity, boot success, etc.

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high-fast)*

Author: Fixed into next commit

- **[CNV-77501]** — As a VM owner, I want to migrate storage for offline VMs between nodes using the same storage class (HPP)
  - *Test Scenario:* [Tier 2] Verify storage migration completes for offline VM from HPP to HPP storage class across different nodes, enabling node-to-node migration without requiring RWX storage
  - *Priority:* P1

- **[CNV-77501]** — As a VM owner, I want an offline VM to still point to the original volume when migration fails
  - *Test Scenario:* [Tier 2] Verify an offline VM still points to the original volume when migration fails
  - *Priority:* P2

- **[CNV-77501]** — As a VM owner, I want the migration to succeed when starting a stopped VM during migration
  - *Test Scenario:* [Tier 2] Verify migration succeeds when starting a VM during the migration process and the VM waits for migration completion before becoming ready
  - *Priority:* P2

---

### **IV. Sign-off and Approval**

This Software Test Plan requires approval from the following stakeholders:

* **Reviewers:**
  - QE Architect (OCP-V): Ruth Netser (`@rnetser`)
  - QE Members (OCP-V): Jenia Peimer (`@jpeimer`), Kate Shvaika (`@kshvaika`), Jose Manuel Castano (`@joscasta`)

> * **Approvers:**

Reviewer: **[HIGH] Approvers: use required Dev Lead / PM role labels**

AGENTS.md requires explicit role labels (QE Lead, Dev Lead, PM). Please relabel Dev Lead and PM as follows (keep the existing QE Architect reviewer/approver entry as-is):

```suggestion
* **Approvers:**
  - QE Architect (OCP-V): Ruth Netser (`@rnetser`)
  - Dev Lead: Alexander Wels (`@awels`)
  - PM: Peter Lauterbach (`@pelauter`)
```

---
*Assisted-by: PI (cursor:cursor-grok-4.5-high-fast)*

Author: Fixed into next commit

### **II. Software Test Plan (STP)**

This STP serves as the **overall roadmap for testing**, detailing the scope, approach, resources, and schedule.

#### **1. Scope of Testing**

**Testing Goals**

- **[P0]** Verify offline VM storage migration completes between ODF and HPP storage classes, and the VM boots successfully after migration
- **[P0]** Verify storage migration completes for a migration plan containing both offline and running VMs
- **[P0]** Verify source volumes are retained or deleted according to retentionPolicy configuration when offline VM storage migration completes
- **[P1]** Verify offline VM storage migration completes when the VM has hotplug disks attached
- **[P2]** Verify offline VM continues pointing to the original volume when storage migration fails
- **[P2]** Verify storage migration completes when a stopped VM is started during the migration process and the VM waits for migration completion before becoming ready

**Storage Class Coverage**

The following storage classes are supported for migration testing:
- **ODF** (ocs-storagecluster-ceph-rbd-virtualization)
- **HPP** (hostpath-csi-pvc-block)
- **AWS EBS** (gp3-csi)
- **Azure Disk** (managed-csi)

> - **GCP PD** (pd-ssd-csi)

Reviewer 3: We will not be testing with these storage classes

Author: Done in next commit

Migration combinations will be tested including:
- Cross-storage class migrations (e.g., ODF ↔ HPP, ODF ↔ AWS EBS)

> - Same-storage class migrations (e.g., ODF ↔ ODF, HPP ↔ HPP)

Reviewer 3: No `AWS EBS`

Author: Removed in next commit

  - *Details:* No blocking dependencies

- [x] **Cross Integrations** — Does the feature affect other features or require testing by other teams? Identify the impact we cause.
  - *Details:* UI team needs to update their migration UI to support offline VM selection

**Infrastructure**

- [x] **Cloud Testing** — Does the feature require multi-cloud platform testing? Consider cloud-specific features.
  - *Details:* Multi-cloud platform testing is required to validate storage migration across cloud provider storage classes (AWS EBS, Azure Disk, GCP PD). Since CDI copy clone performs the underlying storage transfer, the feature works across all platforms.

#### **3. Test Environment**

- **Cluster Topology:** 3-master/3-worker

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization 4.22

- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM

- **Special Hardware:** N/A

- **Storage:**
  - Bare Metal: ocs-storagecluster-ceph-rbd-virtualization, hostpath-csi-pvc-block
  - AWS: gp3-csi
  - Azure: managed-csi

>   - GCP: pd-ssd-csi

Reviewer 3: Please remove these

Author: Done in next commit

# Openshift-virtualization-tests Test plan

## **Offline Storage Migration - Quality Engineering Plan**

### **Metadata & Tracking**

> - **Enhancement(s):** [CNV-77501](https://redhat.atlassian.net/browse/CNV-77501) - no VEP exists for this feature

Reviewer: Minor: Jira links use the old `redhat.atlassian.net` URL format. Please update to `issues.redhat.com` (e.g., `https://issues.redhat.com/browse/CNV-77501`). Applies to Enhancement, Feature Tracking, and Epic Tracking links.

Author: Updated all Jira links in next commit

Reviewer: Verified fix in code — all Jira links updated to issues.redhat.com format. Resolving.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

    - **Documentation:** User guide updates for offline VM storage migration support
    - **UI:** UI updates for offline VM migration selection (covered by UI team in CNV-77503)
  - *Note any NFRs not covered and why:*
    - **Performance:** Not covered - Performance testing for bulk offline migrations is tracked separately and will be addressed by a dedicated test plan
    - **Monitoring:** Not applicable - Feature does not introduce new metrics or alerts; existing migration monitoring applies
    - **Observability:** Not applicable - Feature reuses existing migration observability patterns without new requirements
    - **Security:** Not applicable - Feature does not introduce new security boundaries or authentication/authorization requirements; leverages existing migration RBAC
    - **Scalability:** Inherits existing limits - Feature reuses existing migration infrastructure with its cluster-level parallelism constraints; offline VM support introduces no additional scalability requirements

#### **2. Known Limitations**

The limitations are documented to ensure alignment between development, QA, and product teams.
The following topics will not be tested or supported.

> None - reviewed and confirmed with Jose Manuel Castano on Apr 28,2026.

Reviewer: Known Limitations and Test Limitations are both signed by the QE Owner / PR author. Sign-off on limitations should come from PM or Engineering Manager — not the person writing their own assessment. Please update to a PM or Lead sign-off (e.g., Peter Lauterbach, same as the Out of Scope PM sign-off).

Author: Updaetd both sign-offs, Known limitations and Test limitations into next commit

Reviewer: Verified fix in code — sign-offs updated to PM (Peter Lauterbach). Resolving.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

  - *Note any requirements that are unclear or untestable:* Requirements are testable. Downstream build with the feature code is available for testing.

- [x] **Acceptance Criteria**
  - *List the acceptance criteria:*
    - As a VM owner, I want to migrate storage for offline VMs between ODF and HPP storage classes, so that:
      - Migration plan status reports "Succeeded"
      - Offline VM disk references point to the new target storage class
      - VM boots successfully after migration using the migrated storage
    - As a VM owner, I want to migrate storage with mixed offline VMs and running VMs in one migration plan, so that:
      - Migration plan status reports "Succeeded"
      - All offline VMs point to target storage class
      - All running VMs point to target storage class and remain running
    - As a VM owner, I want to migrate storage for offline VMs with hotplug disk, so that:
      - Migration plan status reports "Succeeded"
      - All disks including hotplug disks are migrated to target storage class
      - VM boots successfully with all disks accessible
    - As a VM owner, I want to retain or delete the source volume for an offline VM, so that:

>       - When retentionPolicy=retain: source volumes exist after migration completes

Reviewer: **[WARNING] `retentionPolicy` implementation term still present in acceptance criteria, testing goal, and test scenario**

The requirement line (40) was correctly rewritten to user-facing language ("source volume cleanup policy"), but the API field name `retentionPolicy` still appears in:
- Acceptance criteria (lines 69-70): `retentionPolicy=retain` / `retentionPolicy=delete`
- Testing goal (line 136): "according to retentionPolicy configuration"
- Test scenario (line 328): "when retentionPolicy is set"

Suggestion: Replace with user-facing language throughout. For example:
- "When the cleanup policy is set to retain: source volumes exist after migration completes"
- "When the cleanup policy is set to delete: source volumes are deleted after migration completes"
- Testing goal: "according to the source volume cleanup policy"
- Test scenario: "when the source volume cleanup policy is configured"

Author: Fixed in next commit

Reviewer 3: I don't think that removing the `retentionPolicy` adds value to this STP, but if that's the requirement, then okay.

Author: Seems is only to follow conventions, I have your same opinion

- **[P0]** Verify storage migration completes for a migration plan containing both offline and running VMs
- **[P0]** Verify source volumes are retained or deleted according to retentionPolicy configuration when offline VM storage migration completes
- **[P1]** Verify offline VM storage migration completes when the VM has hotplug disks attached
- **[P1]** Verify offline VM storage migration completes for same-storage class migrations (HPP to HPP) to enable node-to-node migration with local storage
- **[P2]** Verify offline VM continues pointing to the original volume when storage migration fails
- **[P2]** Verify storage migration completes when a stopped VM is started during the migration process and the VM waits for migration completion before becoming ready

**Storage Class Coverage**

The following storage classes are supported for migration testing:
- **ODF** (ocs-storagecluster-ceph-rbd-virtualization)
- **HPP** (hostpath-csi-pvc-block)

Migration combinations will be tested:
- Cross-storage class migrations (ODF ↔ HPP)
- Same-storage class migrations (HPP → HPP) for node-to-node migration scenarios with local storage

> **Volume Mode Coverage**

Reviewer: **[CRITICAL] Volume mode coverage documented but not traceable to any test scenario**

The requirement (line 42) states "Support all volume mode combinations for cross-storage migration (File-to-Block, Block-to-File, File-to-File, Block-to-Block)" and this section (lines 152-158) commits to testing all 4 combinations. However, none of the 7 test scenarios in Section III mention volume mode at all — they all use generic "between ODF and HPP" without specifying which volume mode is being tested.

Suggestion: Expand the existing offline VM migration scenario (Scenario 1) to explicitly state it covers all volume mode combinations as a test parameter. For example:

> *Test Scenario:* [Tier 2] Verify storage migration completes for offline VMs between ODF and HPP across all volume mode combinations (Block-to-Block, File-to-File, Block-to-File, File-to-Block), and the VM boots successfully after migration

Author: Done in next commit

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g., status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g., status conditions, events, or notifications indicating success or failure)
  - *Details:* UI testing will be covered in https://redhat.atlassian.net/browse/CNV-77503

- [x] **Monitoring** — Does the feature require metrics and/or alerts?
  - *Details:* No new metrics or alerts are required. Existing migration monitoring (including progress field) applies to both online and offline VM migrations. Progress reporting for offline migrations must be verified to ensure it functions correctly.

**Integration & Compatibility**

- [x] **Compatibility Testing** — Ensures feature works across supported platforms, versions, and configurations

>   - Does the feature maintain backward compatibility with previous API versions and configurations?

Reviewer: **[WARNING] Self-Validation Testing type missing from Test Strategy**

AGENTS.md requires all 14 testing types from the template to be present in Section II.2. This STP has 13/14 — "Self-Validation Testing" is absent. Per guidelines: "Unchecked items without details = incomplete review."

Suggestion: Add after Regression Testing:

```markdown
- [ ] **Self-Validation Testing** — Should any of the new tests be included in the self-validation test package?
  - *Details:* N/A. Offline storage migration is not a core operational health-check scenario suitable for the self-validation package.
```

Author: Added in next commit

- [ ] **Upgrade Testing** — Validates upgrade paths from previous versions, data migration, and configuration preservation
  - *Details:* Upgrade path was evaluated. No dedicated upgrade testing is required because this is a new additive feature (offline VM storage migration support) that does not modify existing migration behavior or introduce configuration changes requiring migration. Existing online VM migration functionality remains unchanged and backward compatible (see Compatibility Testing). Users upgrading to v4.22+ will gain offline migration capability without requiring any configuration updates or data migration.

- [ ] **Dependencies** — Blocked by deliverables from other components/products. Identify what we need from other teams before we can test.
  - *Details:* No blocking dependencies

- [x] **Cross Integrations** — Does the feature affect other features or require testing by other teams? Identify the impact we cause.
  - *Details:* UI team needs to update their migration UI to support offline VM selection

**Infrastructure**

> - [x] **Cloud Testing** — Does the feature require multi-cloud platform testing? Consider cloud-specific features.

Reviewer: **[WARNING] Cloud Testing checked but inconsistent with test environment and test scenarios**

Cloud Testing states multi-cloud platform testing is required and names specific cloud storage classes (AWS EBS, Azure Disk, GCP PD). However:
1. The **Storage field** (line 237) only lists bare-metal storage classes — no cloud storage entries
2. **No cloud-specific test scenarios** exist in Section III despite Cloud Testing being marked applicable

Suggestion: Either (a) add cloud storage class entries to the Storage field and add at least one cloud-specific test scenario, or (b) if cloud storage uses platform defaults and doesn't require dedicated scenarios, clarify that in the Cloud Testing details and adjust the Platform field accordingly.

Author: Added option b) to the STP as cloud uses platform does not require dedicated scenarios

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization 4.22

- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM

- **Special Hardware:** N/A

- **Storage:**
  - Bare Metal: ocs-storagecluster-ceph-rbd-virtualization, hostpath-csi-pvc-block

- **Network:** OVN-Kubernetes, IPv4

- **Required Operators:** N/A

- **Platform:** Bare Metal, AWS, Azure, GCP

- **Special Configurations:** N/A

#### **3.1. Testing Tools & Frameworks**

- **Test Framework:** Standard

- **CI/CD:** N/A

- **Other Tools:** N/A

#### **4. Entry Criteria**

The following conditions must be met before testing can begin:

- [x] Requirements and design documents are **approved and merged**
- [x] Test environment can be **set up and configured** (see Section II.3 - Test Environment)

#### **5. Risks**

**Timeline/Schedule**

- **Risk:** No scheduling or deadline risks identified
  - **Mitigation:** Standard test timeline is sufficient for planned test scenarios
  - *Estimated impact on schedule:* None
  - *Sign-off:* Jose Manuel Castano, Apr 28, 2026

**Test Coverage**

> - **Risk:** No gaps in test coverage identified

Reviewer: **[WARNING] Risk section format: "no risk" entries include unnecessary structure; remove "Other" category**

Per AGENTS.md Section II.5 Risks:
- "When no risk exists: **only** a short justification in the Mitigation field is needed; no Sign-off or category-specific supplemental fields are required"
- "'Other' category is included **only** if risks exist that don't fit the 6 standard categories"

All entries currently describe "no risk" but include full Sign-off and supplemental fields. The "Other" section (lines 306-309) has no actual risk ("No additional risks identified") and should be removed entirely.

Suggestion for a "no risk" entry format:
```markdown
**Timeline/Schedule**

- **Risk:** No scheduling or deadline risks identified
  - **Mitigation:** Standard test timeline is sufficient for planned test scenarios
```

Remove the "Other" section completely.

Author: Removed in next commit

  - *Test Scenario:* [Tier 2] Verify storage migration for offline VM from HPP to HPP storage class across different nodes: migration plan status reports "Succeeded", VM disk references point to the new volume on the target node, and VM boots successfully after migration
  - *Priority:* P1

- **[CNV-77501]** — As a VM owner, I want an offline VM to still point to the original volume when migration fails
  - *Test Scenario:* [Tier 2] Verify offline VM rollback on migration failure: migration plan status reports "Failed", and VM disk references remain unchanged pointing to the original storage
  - *Priority:* P2

- **[CNV-77501]** — As a VM owner, I want the migration to succeed when starting a stopped VM during migration
  - *Test Scenario:* [Tier 2] Verify migration succeeds when starting a stopped VM during migration: migration plan status reports "Succeeded", VM remains in a pending state throughout the migration, VM becomes ready only after migration completes, and VM points to the target storage class
  - *Priority:* P2

- **[CNV-77501]** — As a VM owner, I want to cancel an in-progress storage migration for offline and online VMs

>   - *Test Scenario:* [Tier 2] Verify storage migration cancellation while the migration is actively in progress for both offline and online VMs: cancel the migration, verify VM disk references remain unchanged pointing to the original storage, online VMs remain running throughout, and source volumes are preserved

Reviewer: The AC (L85) promises source volumes are preserved *regardless of the configured cleanup policy* — the meaningful case being cleanup policy = **delete** (cancel must not delete source volumes since migration did not complete). This goal (L152) and scenario say only "source volumes are preserved" without naming the retain-vs-delete dimension. Name it explicitly and test cancellation under **both** cleanup policies (especially delete), per AGENTS.md § II.1 (name all config dimensions) / § III (traceable to AC).

---
*Assisted-by: PI (claude-opus-4-8)*

Author: Cleanup policy is not a testing goal for the cancellation scenario and it will just use the default `keepSource`. Let me be more explicit about this in next commit.

>   - *Test Scenario:* [Tier 2] Verify storage migration cancellation while the migration is actively in progress for both offline and online VMs: cancel the migration, verify VM disk references remain unchanged pointing to the original storage, online VMs remain running throughout, and source volumes are preserved

Reviewer: The AC (L82) says the primary cancellation outcome is "the migration stops processing," but this scenario only checks after-effects (disk refs, VM running, source volumes). Add an observable assertion that the migration actually reached the cancelled state — e.g., migration plan status reports "Canceled" / progress halts — matching how every sibling scenario asserts plan status ("Succeeded"/"Failed"). Otherwise a silently no-op'd or still-in-flight cancellation is not cleanly distinguished. AGENTS.md § III.

---
*Assisted-by: PI (claude-opus-4-8)*

Author: @awels should be consider adding a "Cancel" status for a migration plan that has been cancel?

Reviewer 2: That is a good idea, could you make a JIRA card for that?

Author: Already opened a jira ticket for this. I am going to update the test scenario to reflect this cancel status in next commit

# Openshift-virtualization-tests Test plan

## **Offline Storage Migration - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** https://github.com/kubevirt/kubevirt-migration-controller/pull/32
- **Feature Tracking:** [CNV-73509](https://redhat.atlassian.net/browse/CNV-73509)
- **Epic Tracking:** [CNV-73500](https://redhat.atlassian.net/browse/CNV-73500)

> - **QE Owner(s):** Yan Du

Author: Thanks! I have changed ownership and added email

- **Owning SIG:** sig-storage
- **Participating SIGs:** sig-storage

**Document Conventions:**
None

### **Feature Overview**

This feature extends the OpenShift Virtualization migration plan to support storage migration for offline (stopped) VMs in addition to existing online (running) VM support. It enables customers to migrate storage for VMs regardless of their running state, allowing mixed migration plans containing both offline and running VMs.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

This section documents the mandatory QE review process. The goal is to understand the feature's value,
technology, and testability before formal test planning.

#### **1. Requirement & User Story Review Checklist**

<!-- **How to complete this checklist:**
1. **Checkbox**: Mark [x] if the check is complete; if the item cannot be checked - add an explanation why in the `details` section
2. Complete the relevant, needed details for the checklist item -->

> 

Author: Let me remove it in next commit

- [x] **Review Requirements**
  <!-- Review the D/S (Downstream) requirements as defined in Jira. Understand the difference between Upstream (U/S) and D/S requirements.
  Example: "VMs must migrate without network downtime exceeding the defined threshold during node maintenance." -->
  - *List the key D/S requirements reviewed:* Reviewed the user cases for offline VM storage migration from CNV-82430 and CNV-73500

- [x] **Understand Value and Customer Use Cases**
  <!-- Understand why the feature matters to customers from a D/S perspective and what the real-world use cases are.
  Example: "Customers need to migrate VMs without network downtime to maintain SLA compliance during node maintenance." -->
  - *Describe the feature's value to customers:* Enables customers to perform storage migration for offline VMs without requiring them to be running, providing flexibility in storage management operations.

>   - *List the customer use cases identified:* Storage migration for mixed offline VMs and running VMs in one migration plan, allowing batch migration operations regardless of VM state.

Author: Changed to follow more user story format.

>   - *List the customer use cases identified:* Storage migration for mixed offline VMs and running VMs in one migration plan, allowing batch migration operations regardless of VM state.

Reviewer: @josemacassan please do not resolve comments without addressing them
as pointed out by code coeRabbit, this section needs to list customer scenarios

>   - *List the customer use cases identified:* Storage migration for mixed offline VMs and running VMs in one migration plan, allowing batch migration operations regardless of VM state.

Author: Already addressed, keep resolved

  Example: "VM migrates without network downtime exceeding 500ms", "VM deletion is blocked for non-admin users." -->
  - *List the acceptance criteria:*
    - Storage migration completes successfully for offline VMs between ODF and HPP storage classes
    - Storage migration completes successfully for mixed offline VMs and running VMs
    - Storage migration completes successfully for offline VMs with hotplug disk
    - Source volume could be retained/cleaned up for an offline VM migration completed with retentionPolicy defined
    - Offline VM points to the origin volume when migration failed
    - Migration succeeded when starting a stopped VM during migration

  - *Note any gaps or missing criteria:* N/A

- [x] **Non-Functional Requirements (NFRs)**
  <!-- Confirmed coverage for NFRs, including Performance, Security, Usability, Downtime, Connectivity, Monitoring (alerts/metrics), Scalability, Portability (e.g., cloud support), and Docs.-->
  - *List applicable NFRs and their targets:* Documentation updates to reflect offline VM storage migration support

>   - *Note any NFRs not covered and why:* Performance and scale testing are not included in this test plan

Author: Explicitly addressed all NFRs in next commit

#### **2. Known Limitations**

The limitations are documented to ensure alignment between development, QA, and product teams.
The following topics will not be tested or supported.

<!-- Document any known limitations, constraints, or trade-offs in the feature implementation or testing approach.

**Examples:**
Feature related:
- The feature is only supports YYY storage class
- Feature does not support IPv6 (only IPv4)
- No support for ARM64 architecture in this release
- The feature is incompatible with ZZZ feature

Tests related:
- CPU xxx will not be tested due to lack of hardware
- Integration with [Third-Party Service] is excluded; all external calls will be mocked using static data-->

> None identified at this time.

Author: Added the sing-off in next commit

  image registry
- **[P0]** Validate zero data loss during live migration under network latency up to 100ms

**Example - Integration Goals**:
- **[P0]** Verify backward compatibility: upgrade from OCP 4.19 to 4.20 preserves existing VM
  configurations without manual intervention
- **[P0]** Confirm interoperability with OpenShift Service Mesh when VMs use Istio sidecar injection
- **[P1]** Test integration with OpenShift monitoring stack: metrics appear in Prometheus,
  alerts fire correctly in Alertmanager -->

- **[P0]** Verify offline VM storage migration completes between ODF and HPP storage classes, and the VM boots successfully after migration
- **[P0]** Verify storage migration completes for a migration plan containing both offline and running VMs
- **[P1]** Verify offline VM storage migration completes when the VM has hotplug disks attached
- **[P0]** Verify source volumes are retained or deleted according to retentionPolicy configuration when offline VM storage migration completes
- **[P2]** Verify offline VM continues pointing to the original volume when storage migration fails

> - **[P2]** Verify storage migration completes when a stopped VM is started during the migration process

Author: Reordered goals in next commit

**Storage Class Coverage**

The following storage class migration combinations will be tested:

> - **ODF (ocs-storagecluster-ceph-rbd-virtualization) ↔ HPP (hostpath-provisioner)

Author: Done in next commit

  - *Details:* Performance testing for bulk offline migrations is tracked separately in CNV-82430 and will be covered by a separate test plan

- [ ] **Scale Testing** — Validates feature behavior under increased load and at production-like scale (e.g., large number of VMs, nodes, or concurrent operations)
  - *Details:* Not applicable

- [ ] **Security Testing** — Verifies security requirements, RBAC, authentication, authorization, and vulnerability scanning
  - *Details:* Not applicable

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g., status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g., status conditions, events, or notifications indicating success or failure)
  - *Details:* UI testing will be covered in https://redhat.atlassian.net/browse/CNV-77503

> 

Author: Is explained but main development belongs to UI Team, even if we cover in jira described in details section. We wan to leave it as it is.

- [ ] **Cloud Testing** — Does the feature require multi-cloud platform testing? Consider cloud-specific features.
  - *Details:* Not applicable

#### **3. Test Environment**

<!-- **Note:** "N/A" means explicitly not applicable. All items must be filled or marked N/A. -->

- **Cluster Topology:** 3-master/3-worker bare-metal
  <!-- Change if different, e.g., SNO, Compact Cluster, HCP -->

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization 4.22
  <!-- Specify exact versions to allow version traceability -->

- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled
  <!-- Change only if specific CPU requirements exist -->

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM
  <!-- Adjust based on feature requirements -->

- **Special Hardware:** N/A
  <!-- Fill if needed, e.g., SR-IOV NICs, GPUs -->

- **Storage:** ocs-storagecluster-ceph-rbd-virtualization, hostpath-provisioner
  <!-- Change if specific StorageClass(es) required -->

- **Network:** OVN-Kubernetes, IPv4
  <!-- Change if needed, e.g., Secondary Networks, IPv6, dual-stack -->

- **Required Operators:** N/A
  <!-- Add if needed, e.g., NMState Operator -->

> - **Platform:** PSI

Author: Is specify as a concrete platform as PSI is a type of platform

- [ ] All planned test cycles completed
- [ ] Test summary report approved
- [ ] Acceptance criteria met

#### **6. Risks**

<!-- Document specific risks for this feature. If a risk category is not applicable, mark as "N/A" with
justification in mitigation strategy. -->

**Timeline/Schedule**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Estimated impact on schedule:* None

**Test Coverage**

- **Risk:** N/A
  - **Mitigation:** All acceptance criteria are covered by planned test scenarios
  - *Areas with reduced coverage:* None

**Test Environment**

- **Risk:** N/A
  - **Mitigation:** Standard test environment is sufficient for testing this feature
  - *Missing resources or infrastructure:* None

**Untestable Aspects**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Alternative validation approach:* N/A

**Resource Constraints**

- **Risk:** N/A
  - **Mitigation:** N/A
  - *Current capacity gaps:* None

**Dependencies**

- **Risk:** N/A
  - **Mitigation:** No external dependencies
  - *Dependent teams or components:* UI team for UI updates (non-blocking)

**Other**

- **Risk:** N/A
  - **Mitigation:** No additional risks identified

>   <!-- If more context is needed for this item, add an entry with any relevant details-->

Author: In this case is not applicable in any case and the ones that have a mitigation description is already added

>   <!-- If more context is needed for this item, add an entry with any relevant details-->

Author: OK, done in next commit

---

### **III. Test Scenarios & Traceability**

<!-- This section links D/S requirements to test coverage, enabling reviewers to verify all requirements are tested.

**What goes here:** New test scenarios specific to this feature. Each scenario should trace to a D/S Jira requirement.
**Granularity:** If one scenario can fail while another passes, they should be separate items. For example:
- Different conditions (VMs with feature X vs without) → separate scenarios
- Feature Gate enable/disable → separate scenarios (different expected behavior)
- Multiple alerts → separate if different trigger conditions, group if same flow
**What does NOT go here:** Regression tests (covered in Test Strategy). -->

<!-- **Requirement ID:**
- Use Jira issue key (e.g., CNV-12345)

**Requirement Summary:** Brief description from the Jira issue (user story format preferred) -->

- **[CNV-73500]** — As a VM owner, I want to migrate storage for offline VMs between ODF and HPP

>   - *Test Scenario:* [Tier 2] Verify storage migration completes for offline VMs between ODF and HPP, and the VM boots sucessfully after migration

Author: Done in next commit!
